llama-rs/src/llama.rs

```diff
@@ -575,55 +629,56 @@ impl LlamaModel {
         Ok((model, vocab))
     }
 
-    pub fn inference_with_prompt(
-        &self,
-        vocab: &GptVocab,
+    pub fn inference_with_prompt<'a>(
+        &mut self,
+        vocab: &'a GptVocab,
         params: &InferenceParams,
         prompt: &str,
         rng: &mut impl rand::Rng,
+        stop_after_prompt: bool,
+        callback: impl Fn(OutputToken<'a>),
```

Comment:
Instead of adding a new parameter to this function would it be better to break this out into 2 separate functions.  One to feed the model prompt data and one to run an inference.   A bool here doesn't feel like the best API to me.  The model would then need to track its own last_n_tokens and maybe more but that doesn't seem unreasonable to me.

We might then be able to get away from the callback and just have the inference return the tokens, giving the user more control over when to run inferencing instead of forcing it to run to completion.  Putting the core loop back into user code.

Reply:
What I was considering was exposing more of the raw inference loop, but wrapping it up in APIs that make it straightforward to use, so that people can choose the behaviour they need. Not sure what that API would look like yet (would need to look at it after work), but I would expect it to enable prompt caching and continued inference.

Reply:
That sounds great!  I'm coming from the Bevy world so a good API in my opinion gives short running functions that does the minimal unit of work and then gives control back to the caller.  The callback system is inverted to how I would want to use this in projects.

Reply:
Yeah, I added the callback as the smallest-change-possible solution to the existing mechanism. I don't want to expose too much of the details around actually driving the model, but I would like to enable users to dig around.

I think the easiest way to experiment with this is to replicate the loop in user code (I think all of the relevant functions are public) and seeing what we can hide behind the scenes while still allowing for flexibility. Suggestions welcome!

Reply:
How about something like this: A `Model::feed_prompt` method you can use to feed the model a sentence. This would not perform any inference, just the prompt.

Then, another method `Model::next_token` that given the current state of the model memory, returns the next token with no external input. You can call this as many times as you want on a loop, react to the results, and stop whenever you want.

Finally, we can keep the original `infer_with_prompt` as a trivial wrapper over the two previous methods.

This new structure would help a lot in untangling the current mess that `infer_with_prompt` is, and I don't see many downsides to it.

What do you think?

Reply:
Seems reasonable to me! That also solves a problem I've had with llamacord - there's no way to stop generation when it starts. Having manual control over fetching the next token should sort that out.

Would you be able to feed a prompt after calling `next_token`?

Reply:
Yes, feed prompt or next token could be interleaved in any way the user wants. 😄

Reply:
>  Are you planning on building it?

The API? Yes :grin: It's up already!

I had no direct plans, but I'd be super interested in seeing ReAct working with LLaMA using this :eyes: